File: scripts/migrate_turso_tickets.py
"""
migrate_turso_tickets.py
Đọc lark_tickets từ Turso → upsert vào Supabase lark_cs_tickets

Setup:
  Set env vars (hoặc tạo .env.local):
    TURSO_URL=libsql://...
    TURSO_AUTH_TOKEN=eyJ...
    SUPABASE_URL=https://...supabase.co
    SUPABASE_SERVICE_KEY=eyJ...

Chạy:
  python scripts/migrate_turso_tickets.py
  # hoặc:
  C:/Users/hieuh/AppData/Local/Programs/Python/Python311/python.exe scripts/migrate_turso_tickets.py
"""

import json, os
import requests
from pathlib import Path

# ── Load env from .env.local nếu có ──────────────────────────────────────────
env_file = Path(__file__).parent.parent / "web" / ".env.local"
if env_file.exists():
    for line in env_file.read_text(encoding="utf-8").splitlines():
        line = line.strip()
        if line and not line.startswith("#") and "=" in line:
            k, _, v = line.partition("=")
            os.environ.setdefault(k.strip(), v.strip())

TURSO_URL    = os.environ.get("TURSO_URL", "")
TURSO_TOKEN  = os.environ.get("TURSO_AUTH_TOKEN", "")
SUPABASE_URL = os.environ.get("NEXT_PUBLIC_SUPABASE_URL", "")
SUPABASE_KEY = os.environ.get("SUPABASE_SERVICE_KEY", "")

if not TURSO_URL or not TURSO_TOKEN:
    raise SystemExit("Thieu TURSO_URL hoac TURSO_AUTH_TOKEN trong env vars")
if not SUPABASE_URL or not SUPABASE_KEY:
    raise SystemExit("Thieu NEXT_PUBLIC_SUPABASE_URL hoac SUPABASE_SERVICE_KEY trong env vars")

# Convert libsql:// → https://
TURSO_HTTP = TURSO_URL.replace("libsql://", "https://").replace("http://", "https://")

# ── HTTP helpers ──────────────────────────────────────────────────────────────
# 1 Session dùng chung → HTTP Keep-Alive, tái sử dụng kết nối TLS cho mọi batch
# (120+ batch cho 24k+ dòng) thay vì bắt tay TLS mới mỗi request như urllib.
session = requests.Session()
session.verify = False   # giữ nguyên hành vi bản cũ (ssl.CERT_NONE)
try:
    from urllib3.exceptions import InsecureRequestWarning
    requests.packages.urllib3.disable_warnings(InsecureRequestWarning)
except Exception:
    pass

def http_post(url, headers, body_dict):
    res = session.post(url, json=body_dict, headers=headers)
    res.raise_for_status()
    return res.json()

def turso_query(sql):
    body = {"requests": [{"type": "execute", "stmt": {"sql": sql}}, {"type": "close"}]}
    headers = {"Content-Type": "application/json", "Authorization": f"Bearer {TURSO_TOKEN}"}
    resp   = http_post(f"{TURSO_HTTP}/v2/pipeline", headers, body)
    result = resp["results"][0]
    if result["type"] != "ok":
        raise Exception(f"Turso error: {result}")
    cols = [c["name"] for c in result["response"]["result"]["cols"]]
    rows = result["response"]["result"]["rows"]
    return [dict(zip(cols, [cell.get("value") if isinstance(cell, dict) else cell for cell in row])) for row in rows]

def supabase_upsert(rows):
    url     = f"{SUPABASE_URL}/rest/v1/lark_cs_tickets?on_conflict=lark_record_id"
    headers = {
        "Content-Type": "application/json",
        "apikey":        SUPABASE_KEY,
        "Authorization": f"Bearer {SUPABASE_KEY}",
        "Prefer":        "resolution=merge-duplicates",
    }
    res = session.post(url, json=rows, headers=headers)
    if res.status_code >= 400:
        raise Exception(f"Supabase error {res.status_code}: {res.text}")
    return res.status_code

# ── Main ──────────────────────────────────────────────────────────────────────
def main():
    print("[1] Ket noi Turso...")
    count_rows = turso_query("SELECT COUNT(*) as cnt FROM lark_tickets")
    total      = int(count_rows[0].get("cnt", 0))
    print(f"[2] Tong lark_tickets trong Turso: {total}")

    if total == 0:
        print("[!] Turso khong co tickets. Thoat.")
        return

    offset, batch_size, total_upserted = 0, 200, 0
    while offset < total:
        rows = turso_query(
            f"SELECT * FROM lark_tickets ORDER BY creation_date DESC LIMIT {batch_size} OFFSET {offset}"
        )
        if not rows:
            break
        upsert_rows = []
        for r in rows:
            rid = str(r.get("lark_record_id") or "").strip()
            if not rid:
                continue
            cd = r.get("creation_date")
            lt = r.get("leadtime_minutes")
            upsert_rows.append({
                "lark_record_id":   rid,
                "ticket_no":        str(r.get("ticket_no")    or ""),
                "order_no":         str(r.get("order_no")     or ""),
                "sku":              str(r.get("sku")           or ""),
                "ticket_type":      str(r.get("ticket_type")  or ""),
                "issue_detail":     str(r.get("issue_detail") or ""),
                "details":          str(r.get("details")      or ""),
                "source":           str(r.get("source")       or ""),
                "channel":          str(r.get("channel")      or ""),
                "vendor":           str(r.get("vendor")       or ""),
                "handler":          str(r.get("handler")      or ""),
                "product_action":   str(r.get("product_action") or ""),
                "money_action":     str(r.get("money_action")   or ""),
                "creation_date":    int(cd) if cd is not None else None,
                "ticket_status":    str(r.get("ticket_status") or ""),
                "leadtime_minutes": float(lt) if lt is not None else None,
                "source_1":         str(r.get("source_1")    or ""),
            })
        if upsert_rows:
            supabase_upsert(upsert_rows)
        total_upserted += len(upsert_rows)
        offset += batch_size
        print(f"  OK {total_upserted}/{total} tickets upserted...")

    print(f"\nDONE! {total_upserted} tickets migrated to Supabase lark_cs_tickets.")

if __name__ == "__main__":
    main()

File: scripts/migrate_turso_tickets.mjs
/**
 * migrate_turso_tickets.mjs
 * Đọc lark_tickets từ Turso → upsert vào Supabase lark_cs_tickets
 *
 * Chạy: node scripts/migrate_turso_tickets.mjs
 */

import { createClient } from "https://esm.sh/@supabase/supabase-js@2"

const TURSO_URL  = "https://gohub-intel-baole.aws-ap-northeast-1.turso.io"
const TURSO_TOKEN = "eyJhbGciOiJFZERTQSIsInR5cCI6IkpXVCJ9.eyJhIjoicnciLCJpYXQiOjE3NzU0NjMwNzEsImlkIjoiMDE5ZDYxZDctMjcwMS03MDI1LWE0ZGItYmY5NDc0N2Q2ZDNmIiwicmlkIjoiYzlhY2Y3YjYtNTg4OC00Y2QyLWI4ZjktZDA0ODJkYTBhOTg3In0.ReIz6_MHX59QbwMqrlkrCD3NpXZX8aQVDlIjFMFxM29VJ5rR2HLdVWVYat8lLa2AtrkJNEgtka9Y_XtZ-n2HCg"

const SUPABASE_URL = "https://wfuigmfnfcijkvylrwzz.supabase.co"
const SUPABASE_KEY = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6IndmdWlnbWZuZmNpamt2eWxyd3p6Iiwicm9sZSI6InNlcnZpY2Vfcm9sZSIsImlhdCI6MTc4MDQ3MDgxNiwiZXhwIjoyMDk2MDQ2ODE2fQ.0XvbKXseTFTx9iQzmYlXSwoT0bxu2CI8-wrCEMv8ipo"

async function tursoQuery(sql, args = []) {
  const res = await fetch(`${TURSO_URL}/v2/pipeline`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      "Authorization": `Bearer ${TURSO_TOKEN}`,
    },
    body: JSON.stringify({
      requests: [
        { type: "execute", stmt: { sql, args: args.map(v => ({ type: "text", value: String(v ?? "") })) } },
        { type: "close" },
      ],
    }),
  })

  if (!res.ok) throw new Error(`Turso HTTP ${res.status}: ${await res.text()}`)
  const body = await res.json()
  const result = body.results?.[0]
  if (result?.type !== "ok") throw new Error(`Turso error: ${JSON.stringify(result)}`)

  const cols = result.response.result.cols.map(c => c.name)
  const rows = result.response.result.rows
  return rows.map(row =>
    Object.fromEntries(cols.map((col, i) => [col, row[i]?.value ?? row[i] ?? null]))
  )
}

async function main() {
  const supabase = createClient(SUPABASE_URL, SUPABASE_KEY)

  console.log("🔌 Kết nối Turso...")
  const countRows = await tursoQuery("SELECT COUNT(*) as cnt FROM lark_tickets")
  const total = parseInt(String(countRows[0]?.cnt || 0))
  console.log(`📊 Tổng lark_tickets trong Turso: ${total}`)

  if (total === 0) {
    console.log("❌ Turso không có tickets. Thoát.")
    return
  }

  let offset = 0
  const batchSize = 200
  let totalUpserted = 0

  while (offset < total) {
    const rows = await tursoQuery(
      `SELECT * FROM lark_tickets ORDER BY creation_date DESC LIMIT ${batchSize} OFFSET ${offset}`
    )
    if (!rows.length) break

    const upsertRows = rows
      .map(r => ({
        lark_record_id:   String(r.lark_record_id || ""),
        ticket_no:        String(r.ticket_no    || ""),
        order_no:         String(r.order_no     || ""),
        sku:              String(r.sku           || ""),
        ticket_type:      String(r.ticket_type  || ""),
        issue_detail:     String(r.issue_detail || ""),
        details:          String(r.details      || ""),
        source:           String(r.source       || ""),
        channel:          String(r.channel      || ""),
        vendor:           String(r.vendor       || ""),
        handler:          String(r.handler      || ""),
        product_action:   String(r.product_action || ""),
        money_action:     String(r.money_action   || ""),
        creation_date:    r.creation_date ? Number(r.creation_date) : null,
        ticket_status:    String(r.ticket_status || ""),
        leadtime_minutes: r.leadtime_minutes ? Number(r.leadtime_minutes) : null,
        source_1:         String(r.source_1    || ""),
        updated_at:       new Date().toISOString(),
      }))
      .filter(r => r.lark_record_id)

    const { error } = await supabase
      .from("lark_cs_tickets")
      .upsert(upsertRows, { onConflict: "lark_record_id", ignoreDuplicates: false })

    if (error) {
      console.error(`❌ Supabase upsert error (offset ${offset}):`, error.message)
      process.exit(1)
    }

    totalUpserted += upsertRows.length
    offset += batchSize
    console.log(`  ✓ ${totalUpserted}/${total} tickets upserted...`)
  }

  console.log(`\n✅ Hoàn thành! ${totalUpserted} tickets đã migrate sang Supabase lark_cs_tickets.`)
}

main().catch(err => {
  console.error("❌ Error:", err.message)
  process.exit(1)
})
